### PR TITLE
Allow value in ICAs

### PR DESCRIPTION
**Pull request by yorhodes**
### Description

- Allows ICAs to send and receive native value
- Makes ICAs and IQs compatible with non EVM domains

### Drive-by changes

- Unifies ICA and IQS message encoding for interchain calls
- Adds 2% tolerance to gas snapshot checks for velocity

### Related issues

- Fixes https://github.com/hyperlane-xyz/hyperlane-monorepo/issues/1623

### Backward compatibility

No

### Testing

- [x] add ICA unit tests for receiving and sending value

### CODE REVIEW

File: solidity/contracts/middleware/InterchainAccountRouter.sol [hunk 1]
```diff
@@ -90,7 +107,18 @@ contract InterchainAccountRouter is Router, IInterchainAccountRouter {
     ) external returns (bytes32) {
         CallLib.Call[] memory calls = new CallLib.Call[](1);
         calls[0] = CallLib.Call({to: target, data: data});
-        return _dispatch(_destinationDomain, abi.encode(msg.sender, calls));
+        return _dispatch(_destinationDomain, calls.format(msg.sender));
+    }
+
+    function dispatch(
```
Comment (nambrot) [inline solidity/contracts/middleware/InterchainAccountRouter.sol:113]: Optional: I would personally probably just remove this version that doesn't have value as an argument

File: solidity/test/InterchainAccountRouter.t.sol [hunk 2]
```diff
@@ -129,4 +129,30 @@ contract InterchainAccountRouterTest is Test {
         );
         assertEq(localIca.owner(), address(originRouter));
     }
+
+    function testReceiveValue(uint256 value) public {
+        vm.assume(value <= address(this).balance);
```
Comment (nambrot) [inline solidity/test/InterchainAccountRouter.t.sol:134]: Idk foundry enough, but will this test always have a balance?

File: solidity/test/InterchainAccountRouter.t.sol [hunk 3]
```diff
@@ -129,4 +129,30 @@ contract InterchainAccountRouterTest is Test {
...
+        // receive value after deployed
+        remoteRouter.getDeployedInterchainAccount(originDomain, address(this));
+        assert(ica.code.length > 0);
+        ica.transfer(value / 2);
+    }
+
+    // solhint-disable-next-line no-empty-blocks
+    function receiveValue() external payable {}
+
+    function testSendValue(uint256 value) public {
```
Comment (nambrot) [inline solidity/test/InterchainAccountRouter.t.sol:149]: This was a bit confusing too read. IIUC, the ICA makes the call on `this`, and you test that value can be moved. IMO, shoulnd't the test transfer funds to the ICA, and then show that the ICA call can actually be done with value?

File: solidity/contracts/middleware/InterchainAccountRouter.sol (hunk 1 above)
Comment (yorhodes) [reply solidity/contracts/middleware/InterchainAccountRouter.sol:113]: avoiding breaking changes to the interface and gas amounts

File: solidity/test/InterchainAccountRouter.t.sol (hunk 2 above)
Comment (yorhodes) [reply solidity/test/InterchainAccountRouter.t.sol:134]: no

File: solidity/test/InterchainAccountRouter.t.sol (hunk 3 above)
Comment (yorhodes) [reply solidity/test/InterchainAccountRouter.t.sol:149]: your suggestion is what the test is doing
1. ica.transfer(value)
2. dispatch this.receiveValue ICA call that transfers value

File: solidity/contracts/libs/Call.sol [hunk 4]
```diff
@@ -9,43 +9,78 @@ library CallLib {
         bytes data;
     }
 
-    function _multicall(address to, bytes[] memory calls) internal {
+    struct CallWithValue {
```
Comment (asaj) [inline solidity/contracts/libs/Call.sol:12]: Why have a separate type?

File: solidity/contracts/libs/Call.sol [hunk 5]
```diff
@@ -9,43 +9,78 @@ library CallLib {
...
+        uint256 value;
+        Call _call;
+    }
+
+    struct CallWithCallback {
+        Call _call;
+        bytes callbackdata;
```
Comment (asaj) [inline solidity/contracts/libs/Call.sol:19]: Shouldn't `callbackdata` also be a `Call`?

File: solidity/contracts/libs/Call.sol [hunk 6]
```diff
@@ -9,43 +9,78 @@ library CallLib {
         bytes data;
     }
 
-    function _multicall(address to, bytes[] memory calls) internal {
+    struct CallWithValue {
+        uint256 value;
+        Call _call;
```
Comment (asaj) [inline solidity/contracts/libs/Call.sol:14]: The `_` prefix isn't something we usually use in struct fields, is it?

File: solidity/contracts/libs/Call.sol [hunk 7]
```diff
@@ -9,43 +9,78 @@ library CallLib {
...
+    }
+
+    struct CallWithCallback {
+        Call _call;
+        bytes callbackdata;
+    }
+
+    function _call(Call memory call) internal returns (bytes memory) {
```
Comment (asaj) [inline solidity/contracts/libs/Call.sol:22]: Should the functions prefixed with `_` be private? If not, why use it in call but not multicall?

File: solidity/contracts/libs/Call.sol (hunk 5 above)
Comment (asaj) [reply solidity/contracts/libs/Call.sol:19]: Can avoid defining an extra struct (and `call.call`) if you make the callback an arg, e.g.

function call(Call memory call, Call memory cb) internal returns (bytes memory) {}

File: solidity/contracts/libs/Call.sol [hunk 8]
```diff
@@ -9,43 +9,78 @@ library CallLib {
...
                 ++i;
             }
         }
     }
 
-    function _multicallAndResolve(Call[] memory calls, bytes[] memory callbacks)
+    function multicall(CallWithCallback[] memory calls)
         internal
-        returns (bytes[] memory resolveCalls)
+        returns (bytes[] memory callbacks)
```
Comment (asaj) [inline solidity/contracts/libs/Call.sol:66]: It might be less gas efficient but having this return `Call[] memory cbs` would be more readable, IMO

File: solidity/contracts/middleware/InterchainAccountRouter.sol [hunk 9]
```diff
@@ -73,7 +78,19 @@ contract InterchainAccountRouter is Router, IInterchainAccountRouter {
...
-        return _dispatch(_destinationDomain, abi.encode(msg.sender, calls));
+        return _dispatch(_destinationDomain, calls.format(msg.sender));
+    }
+
+    /**
+     * @notice Dispatches a sequence of calls to be relayed by the sender's interchain account on the destination domain.
+     * @param _destinationDomain The domain of the chain where the message will be sent to.
+     * @param calls The sequence of calls to be relayed.
+     */
+    function dispatch(
```
Comment (asaj) [inline solidity/contracts/middleware/InterchainAccountRouter.sol:89]: Four flavors of dispatch feels like a lot, thoughts on just having two, one with CallWithValue[] and the other with CallWithValue?

File: solidity/contracts/libs/Call.sol (hunk 4 above)
Comment (asaj) [reply solidity/contracts/libs/Call.sol:12]: Ah for IQS, ofc. Maybe `Call` and `StaticCall`?

File: solidity/contracts/middleware/InterchainCallMessage.sol [hunk 10]
```diff
@@ -0,0 +1,98 @@
...
+
+import {CallLib} from "../libs/Call.sol";
+
+// TODO: optimize call decoding with calldata slicing
+library InterchainCallMessage {
+    enum Type {
+        DEFAULT,
+        WITH_VALUE,
+        WITH_CALLBACK,
+        RAW_CALLDATA
```
Comment (asaj) [inline solidity/contracts/middleware/InterchainCallMessage.sol:12]: I assume if we encode callbacks as `Calls` we can remove this?

File: solidity/contracts/middleware/InterchainQueryRouter.sol [hunk 11]
```diff
@@ -70,25 +73,34 @@ contract InterchainQueryRouter is Router, IInterchainQueryRouter {
         );
     }
 
+    function query(
```
Comment (asaj) [inline solidity/contracts/middleware/InterchainQueryRouter.sol:76]: Ditto, I think four flavors of query is too many, to the point where the lack of clarity outweighs the optionality

File: solidity/contracts/libs/Call.sol (hunk 5 above)
Comment (yorhodes) [reply solidity/contracts/libs/Call.sol:19]: I actually did this initially
however, callbacks not made on `msg.sender` cannot be allowed and otherwise we would just duplicate `msg.sender`  across the entire array of calls

File: solidity/contracts/libs/Call.sol (hunk 8 above)
Comment (yorhodes) [reply solidity/contracts/libs/Call.sol:66]: agreed more readable

File: solidity/contracts/libs/Call.sol (hunk 5 above)
Comment (yorhodes) [reply solidity/contracts/libs/Call.sol:19]: still need to encode these

File: solidity/contracts/middleware/InterchainAccountRouter.sol (hunk 9 above)
Comment (yorhodes) [reply solidity/contracts/middleware/InterchainAccountRouter.sol:89]: the raw interaction from tools like the explorer motivated this

File: solidity/contracts/libs/Call.sol (hunk 4 above)
Comment (yorhodes) [reply solidity/contracts/libs/Call.sol:12]: `StaticCall` is pretty easy to conflate with `address.staticcall`

File: solidity/contracts/middleware/InterchainAccountRouter.sol (hunk 9 above)
Comment (yorhodes) [reply solidity/contracts/middleware/InterchainAccountRouter.sol:89]: also trying to avoid backwards incompatible changes

File: solidity/contracts/libs/Call.sol (hunk 4 above)
Comment (asaj) [reply solidity/contracts/libs/Call.sol:12]: Exactly, I'm proposing that if you're using the struct without the value field that it's only used for static calls

File: solidity/contracts/libs/Call.sol (hunk 5 above)
Comment (asaj) [reply solidity/contracts/libs/Call.sol:19]: Sorry, are you saying you can't do this?
```
Can avoid defining an extra struct (and call.call) if you make the callback an arg, e.g.

function call(Call memory call, bytes memory cb) internal returns (bytes memory) {}
```

File: solidity/contracts/libs/Call.sol (hunk 4 above)
Comment (asaj) [reply solidity/contracts/libs/Call.sol:12]: i.e.
```
struct StaticCall {
  address to;
  bytes data;
}

struct Call {
  address to;
  bytes data;
  uint256 value;
}

    function _staticcall(StaticCall memory call)
        internal
        view
        returns (bytes memory);

    function _staticcall(StaticCall memory call, bytes memory cb)
        internal
        view
        returns (bytes memory);

    function multicall(StaticCall[] memory calls) internal view returns (bytes[] memory);
    function multicall(StaticCall[] memory calls, bytes[] memory cbs) internal view returns (bytes[] memory);

    
     function _call(Call memory call) internal returns (bytes memory);
     function _call(Call memory call, bytes memory cb) internal returns (bytes memory);
    function multicall(Call[] memory calls) internal returns (bytes[] memory);
    function multicall(Call[] memory calls, bytes[] memory cbs) internal returns (bytes[] memory);
```

**Comment by asaj** [comment]
Hmm this PR should be blocked due to decreased coverage. I'll investigate once you merge main

File: solidity/contracts/libs/Call.sol (hunk 5 above)
Comment (yorhodes) [reply solidity/contracts/libs/Call.sol:19]: we can but then we have to separately encode the arrays which is suboptimal

File: solidity/contracts/libs/Call.sol [hunk 12]
```diff
@@ -4,51 +4,105 @@ pragma solidity ^0.8.13;
...
+    }
+
     struct Call {
         address to;
+        uint256 value;
         bytes data;
     }
 
-    function _multicall(address to, bytes[] memory calls) internal {
+    struct CallWithCallback {
```
Comment (asaj) [inline solidity/contracts/libs/Call.sol:18]: nit: StaticCallWithCallback

File: solidity/contracts/libs/Call.sol [hunk 13]
```diff
@@ -4,51 +4,105 @@ pragma solidity ^0.8.13;
...
-            Address.functionCall(to, calls[i]);
+            call(calls[i]);
             unchecked {
                 ++i;
             }
         }
     }
 
-    function _multicall(Call[] memory calls) internal {
+    function staticmulticall(CallWithCallback[] memory _calls)
```
Comment (asaj) [inline solidity/contracts/libs/Call.sol:57]: nit: multistaticcall? Don't feel strongly

File: solidity/contracts/middleware/InterchainAccountRouter.sol [hunk 14]
```diff
@@ -99,42 +96,59 @@ contract InterchainAccountRouter is Router, IInterchainAccountRouter {
...
      * @return The address of the interchain account.
      */
-    function getDeployedInterchainAccount(uint32 _origin, address _sender)
+    function getDeployedInterchainAccount(uint32 _origin, bytes32 _sender)
         public
         returns (OwnableMulticall)
     {
         bytes32 salt = _salt(_origin, _sender);
-        address interchainAccount = _getInterchainAccount(salt);
+        address payable interchainAccount = _getInterchainAccount(salt);
```
Comment (asaj) [inline solidity/contracts/middleware/InterchainAccountRouter.sol:126]: Use `getInterchainAccount(_origin, _sender)`?

File: solidity/contracts/middleware/InterchainQueryRouter.sol [hunk 15]
```diff
@@ -31,22 +36,22 @@ contract InterchainQueryRouter is Router, IInterchainQueryRouter {
      */
     event QueryDispatched(
         uint32 indexed destinationDomain,
-        address indexed sender
+        bytes32 indexed sender
```
Comment (asaj) [inline solidity/contracts/middleware/InterchainQueryRouter.sol:39]: Should this stay as an address?

File: solidity/contracts/middleware/InterchainQueryRouter.sol [hunk 16]
```diff
@@ -31,22 +36,22 @@ contract InterchainQueryRouter is Router, IInterchainQueryRouter {
...
+    event QueryReturned(uint32 indexed originDomain, bytes32 indexed sender);
     /**
      * @notice Emitted when a query is resolved on the origin chain.
      * @param destinationDomain The domain of the chain that was queried.
      * @param sender The address that resolved the query.
      */
     event QueryResolved(
         uint32 indexed destinationDomain,
-        address indexed sender
+        bytes32 indexed sender
```
Comment (asaj) [inline solidity/contracts/middleware/InterchainQueryRouter.sol:54]: Ditto

File: solidity/contracts/middleware/InterchainQueryRouter.sol [hunk 17]
```diff
@@ -140,27 +97,19 @@ contract InterchainQueryRouter is Router, IInterchainQueryRouter {
...
-                bytes[] memory callbacks
-            ) = abi.decode(
-                    _message,
-                    (Action, address, CallLib.Call[], bytes[])
-                );
-            callbacks = calls._multicallAndResolve(callbacks);
-            _dispatch(_origin, abi.encode(Action.RESOLVE, sender, callbacks));
+        InterchainCallMessage.CallType calltype = _message.calltype();
+        bytes32 sender = _message.sender();
+        if (calltype == InterchainCallMessage.CallType.CALLBACK) {
```
Comment (asaj) [inline solidity/contracts/middleware/InterchainQueryRouter.sol:102]: This is kind of confusing to me, since these are actually the static calls, and the *next* block (raw) is the callbacks.

Comment (asaj) [reply solidity/contracts/middleware/InterchainQueryRouter.sol:102]: A comment could help clarify.

File: solidity/contracts/test/TestQuery.sol [hunk 18]
```diff
@@ -18,12 +19,14 @@ contract TestQuery {
...
-        bytes memory data = abi.encodeWithSignature("owner()");
-        bytes memory callback = bytes.concat(
-            this.receiveRouterOwer.selector,
-            bytes32(secret)
+        CallLib.CallWithCallback[]
+            memory calls = new CallLib.CallWithCallback[](1);
+        calls[0] = CallLib.build(
+            target,
+            abi.encodeWithSignature("owner()"),
+            bytes.concat(this.receiveRouterOwer.selector, bytes32(secret))
```
Comment (asaj) [inline solidity/contracts/test/TestQuery.sol:27]: nit: receiveRouterOwner?

File: solidity/contracts/test/TestQuery.sol [hunk 19]
```diff
@@ -18,12 +19,14 @@ contract TestQuery {
...
         address target = TypeCasts.bytes32ToAddress(router.routers(domain));
-        bytes memory data = abi.encodeWithSignature("owner()");
-        bytes memory callback = bytes.concat(
-            this.receiveRouterOwer.selector,
-            bytes32(secret)
+        CallLib.CallWithCallback[]
+            memory calls = new CallLib.CallWithCallback[](1);
+        calls[0] = CallLib.build(
+            target,
+            abi.encodeWithSignature("owner()"),
```
Comment (asaj) [inline solidity/contracts/test/TestQuery.sol:26]: nit: Could also use `abi.encodeWithSelector()`?

File: solidity/test/InterchainAccountRouter.t.sol [hunk 20]
```diff
@@ -9,12 +9,15 @@ import {TypeCasts} from "../contracts/libs/TypeCasts.sol";
...
 
 contract InterchainAccountRouterTest is Test {
-    // TODO: dedupe
+    using CallLib for address;
+    using TypeCasts for address;
+
     event InterchainAccountCreated(
         uint32 indexed origin,
-        address sender,
+        bytes32 sender,
```
Comment (asaj) [inline solidity/test/InterchainAccountRouter.t.sol:20]: nit: How *do* you dedupe these?

File: solidity/test/InterchainAccountRouter.t.sol (hunk 2 above)
Comment (asaj) [reply solidity/test/InterchainAccountRouter.t.sol:133]: I do :)

File: solidity/test/InterchainAccountRouter.t.sol [hunk 21]
```diff
@@ -129,4 +128,38 @@ contract InterchainAccountRouterTest is Test {
...
+        remoteRouter.getDeployedInterchainAccount(originDomain, address(this));
+        assert(ica.code.length > 0);
+        ica.transfer(value / 2);
+    }
+
+    // solhint-disable-next-line no-empty-blocks
+    function receiveValue() external payable {}
+
+    function testSendValue(uint256 value) public {
+        vm.assume(value <= address(this).balance);
```
Comment (asaj) [inline solidity/test/InterchainAccountRouter.t.sol:149]: Assume value > 0

File: solidity/contracts/middleware/InterchainAccountRouter.sol (hunk 14 above)
Comment (yorhodes) [reply solidity/contracts/middleware/InterchainAccountRouter.sol:126]: salt is used outside the call

File: solidity/test/InterchainAccountRouter.t.sol (hunk 20 above)
Comment (yorhodes) [reply solidity/test/InterchainAccountRouter.t.sol:20]: doesnt seem possible

File: solidity/contracts/OwnableMulticall.sol [hunk 22]
```diff
@@ -9,11 +9,9 @@ import {CallLib} from "./libs/Call.sol";
 
 /*
  * @title OwnableMulticall
- * @dev Allows only only address to execute calls to other contracts
+ * @dev Permits owner address to execute stateful calls with value to other contracts
```
Comment (asaj) [inline solidity/contracts/OwnableMulticall.sol:12]: nit: Consider `state mutating` over statefull

File: solidity/contracts/libs/Call.sol [hunk 23]
```diff
@@ -4,51 +4,109 @@ pragma solidity ^0.8.13;
 import {Address} from "@openzeppelin/contracts/utils/Address.sol";
 
 library CallLib {
+    struct StaticCall {
+        address to;
```
Comment (asaj) [inline solidity/contracts/libs/Call.sol:8]: Can we realistically support non-evm chains if this isn't bytes32? Flagging b/c we switched to bytes32 elsewhere, but is that useful if our messages don't support bytes32 here?

File: solidity/contracts/libs/Call.sol (hunk 5 above)
Comment (asaj) [reply solidity/contracts/libs/Call.sol]: IMO this is cleaner and that may be worth the gas savings but I don't feel strongly

File: solidity/contracts/libs/Call.sol [hunk 24]
```diff
@@ -4,51 +4,109 @@ pragma solidity ^0.8.13;
...
+    }
+
+    function call(Call memory _call)
+        internal
+        returns (bytes memory returnData)
+    {
+        return Address.functionCallWithValue(_call.to, _call.data, _call.value);
+    }
+
+    function staticcall(StaticCall memory _call)
```
Comment (asaj) [inline solidity/contracts/libs/Call.sol:30]: Btw where is this used? Do we ever do static calls without callbacks?

I guess maybe we will later if we implement resolve/reject?

File: solidity/contracts/libs/Call.sol (hunk 23 above)
Comment (asaj) [reply solidity/contracts/libs/Call.sol:8]: Oh I guess you could control ICAs on EVM chains via accounts on non-EVM chains, but not vice versa, is that right?

File: solidity/contracts/middleware/InterchainQueryRouter.sol [hunk 25]
```diff
@@ -38,7 +35,7 @@ contract InterchainQueryRouter is Router, IInterchainQueryRouter {
      * @param originDomain The domain of the chain to return the result to.
      * @param sender The address to receive the result.
      */
-    event QueryReturned(uint32 indexed originDomain, address indexed sender);
+    event QueryReturned(uint32 indexed originDomain, bytes32 indexed sender);
```
Comment (asaj) [inline solidity/contracts/middleware/InterchainQueryRouter.sol:38]: Sorry I'm not sure I understand the difference between QueryReturned and QueryResolved? Returned to the origin chain vs resolved on the origin chain, what's the difference?

Comment (asaj) [reply solidity/contracts/middleware/InterchainQueryRouter.sol:38]: I think maybe the natspec for QueryReturned is wrong?

Comment (asaj) [reply solidity/contracts/middleware/InterchainQueryRouter.sol:38]: QueryExecuted might be a clearer name?

File: solidity/contracts/middleware/InterchainQueryRouter.sol [hunk 26]
```diff
@@ -140,27 +88,27 @@ contract InterchainQueryRouter is Router, IInterchainQueryRouter {
...
+        bytes32 sender = InterchainCallMessage.sender(_message);
+        if (
+            calltype == InterchainCallMessage.CallType.STATIC_CALL_WITH_CALLBACK
+        ) {
             emit QueryReturned(_origin, sender);
-        } else if (action == Action.RESOLVE) {
-            (, address sender, bytes[] memory resolveCallbacks) = abi.decode(
-                _message,
-                (Action, address, bytes[])
+            CallLib.StaticCallWithCallback[]
```
Comment (asaj) [inline solidity/contracts/middleware/InterchainQueryRouter.sol:98]: Should you emit after executing the multistaticcall?

File: solidity/contracts/middleware/InterchainQueryRouter.sol [hunk 27]
```diff
@@ -140,27 +88,27 @@ contract InterchainQueryRouter is Router, IInterchainQueryRouter {
...
+                    .callsWithCallbacks(_message);
+            bytes[] memory callbacks = CallLib.multistaticcall(
+                callsWithCallback
             );
-            sender._multicall(resolveCallbacks);
-            emit QueryResolved(_origin, sender);
+            _dispatch(_origin, InterchainCallMessage.format(callbacks, sender));
+        } else if (calltype == InterchainCallMessage.CallType.RAW_CALLDATA) {
+            address senderAddress = sender.bytes32ToAddress();
+            emit QueryResolved(_origin, senderAddress);
```
Comment (asaj) [inline solidity/contracts/middleware/InterchainQueryRouter.sol:107]: Ditto, shoudl this go after the multicallto?

File: solidity/contracts/middleware/InterchainQueryRouter.sol [hunk 28]
```diff
@@ -140,27 +88,27 @@ contract InterchainQueryRouter is Router, IInterchainQueryRouter {
...
+            bytes[] memory rawCalls = InterchainCallMessage.rawCalls(_message);
+            CallLib.multicallto(rawCalls, senderAddress);
```
Comment (asaj) [inline solidity/contracts/middleware/InterchainQueryRouter.sol:109]: nit: I would expect senderAddress to be the first arg here

File: solidity/package.json [hunk 29]
```diff
@@ -50,7 +50,7 @@
     "test": "hardhat test && yarn test:forge",
     "test:forge": "forge test -vvv",
     "gas": "forge snapshot",
-    "gas-ci": "yarn gas --check || (echo 'Manually update gas snapshot' && exit 1)",
+    "gas-ci": "yarn gas --check --tolerance 2 || (echo 'Manually update gas snapshot' && exit 1)",
```
Comment (asaj) [inline solidity/package.json:53]: Make sure documented in drivebys if not otherwise

File: solidity/contracts/test/TestQuerySender.sol [hunk 30]
```diff
@@ -32,13 +33,13 @@ contract TestQuerySender {
         bytes calldata _targetData,
         uint256 _gasAmount
     ) external payable {
-        bytes32 _messageId = queryRouter.query(
+        queryAndPayFor(
```
Comment (asaj) [inline solidity/contracts/test/TestQuerySender.sol:36]: Make sure documented in drivebys if not otherwise

**Review by asaj — APPROVED** [review]
Mostly nits, please address before merging

File: solidity/contracts/libs/Call.sol (hunk 23 above)
Comment (yorhodes) [reply solidity/contracts/libs/Call.sol:8]: I guess Call.sol is specific to EVM land
we could make InterchainCallMessage encode addresses as bytes32

File: solidity/contracts/libs/Call.sol (hunk 24 above)
Comment (yorhodes) [reply solidity/contracts/libs/Call.sol:30]: making private for now

File: solidity/contracts/middleware/InterchainQueryRouter.sol (hunk 25 above)
Comment (yorhodes) [reply solidity/contracts/middleware/InterchainQueryRouter.sol:38]: > Sorry I'm not sure I understand the difference between QueryReturned and QueryResolved? Returned to the origin chain vs resolved on the origin chain, what's the difference?
are you arguing they should be the same event emitted on each chain?

File: solidity/contracts/middleware/InterchainQueryRouter.sol (hunk 26 above)
Comment (yorhodes) [reply solidity/contracts/middleware/InterchainQueryRouter.sol:98]: if I do this, slither correctly flags that the arbitrary calls out can cause events to be emitted out of order
probably fine but I will have to disable that warning

File: solidity/contracts/test/TestQuerySender.sol (hunk 30 above)
Comment (yorhodes) [reply solidity/contracts/test/TestQuerySender.sol:36]: this is just a test helper?

File: solidity/contracts/middleware/InterchainQueryRouter.sol (hunk 25 above)
Comment (yorhodes) [reply solidity/contracts/middleware/InterchainQueryRouter.sol]: renamed to `QueryExecuted`

File: solidity/contracts/libs/Call.sol (hunk 5 above)
Comment (yorhodes) [reply solidity/contracts/libs/Call.sol]: it is cleaner but imo having the callback be a regular call that can only be on msg.sender is a footgun
